ericw-tools/man/light.1
.\" Process this file with
.\" groff -man -Tascii light.1
.\"
.TH LIGHT 1 "TYR_VERSION" TYRUTILS

.SH NAME
light \- Caclulate lightmap data for a Quake BSP file

.SH SYNOPSIS
\fBlight\fP [OPTION]... BSPFILE

.SH DESCRIPTION
\fBlight\fP reads a Quake .bsp file and calculates light and shadow
information based on the entity definitions contained in the .bsp.  The .bsp
file is updated with the new light data upon completion, overwriting any
existing lighting data.

.SH OPTIONS

.PP
Note, any of the Worldspawn Keys listed in the next
section can be supplied as command-line options, which will override any
setting in worldspawn.
.br
.br

.SS "Performance options:"
.IP "\fB-threads n\fP"
Set number of threads explicitly. By default light will attempt to detect the
number of CPUs/cores available.
.IP "\fB-extra\fP"
Calculate extra samples (2x2) and average the results for smoother shadows.
.IP "\fB-extra4\fP"
Calculate even more samples (4x4) and average the results for smoother
shadows.
.IP "\fB-gate n\fP"
Set a minimum light level, below which can be considered zero brightness.
This can dramatically speed up processing when there are large numbers of
lights with inverse or inverse square falloff. In most cases, values less than
1.0 will cause no discernible visual differences.  Default 0.001.
.IP "\fB-sunsamples [n]\fP"
Set the number of samples to use for "_sunlight_penumbra" and "_sunlight2" (sunlight2 may use more or less because of how the suns are set up in a sphere). Default 100.
.IP "\fB-surflight_subdivide [n]\fP"
Configure spacing of all surface lights. Default 128 units. Minimum setting: 64 / max 2048.
In the future I'd like to make this configurable per-surface-light.
.br
.SS "Output format options:"
.IP "\fB-lit\fP"
Force generation of a .lit file, even if your map does not have any coloured
lights. By default, light will automatically generate the .lit file when
needed.
.IP "\fB-onlyents\fP"
Updates the entities lump in the bsp. You should run this after running qbsp with -onlyents,
if your map uses any switchable lights. All this does is assign style numbers to each
switchable light.
.IP "\fB-litonly\fP"
Generate a .lit file that is compatible with the .bsp without modifying the .bsp.
This is meant for tweaking lighting or adding colored lights when you can't modify
an existing .bsp (e.g. for multiplayer maps.) 
Typically you would make a temporary copy of the .bsp, update the lights in the
entity lump (e.g. with "qbsp -onlyents"), then re-light it with "light -litonly".
Engines may enforce a restriction that you can't make areas brighter than they originally were (cheat protection).
Also, styled lights (flickering/switchable) can't be added in new areas or have their styles changed.

.br
.SS "Postprocessing options:"
.IP "\fB-soft [n]\fP"
Perform post-processing on the lightmap which averages adjacent samples to
smooth shadow edges.  If n is specified, the algorithm will take 'n' samples
on each side of the sample point and replace the original value with the
average. e.g. a value of 1 results in averaging a 3x3 square centred on the
original sample. 2 implies a 5x5 square and so on.  If -soft is specified, but
n is omitted, a value will be the level of oversampling requested. If no
oversampling, then the implied value is 1. -extra implies a value of 2 and
-extra4 implies 3.  Default 0 (off).
.br
.SS "Debug modes:"
.IP "\fB-dirtdebug\fP"
Implies "-dirt", and renders just the dirtmap against a fullbright background,
ignoring all lights in the map. Useful for previewing and turning the dirt settings.
.IP "\fB-phongdebug\fP"
Write normals to lit file for debugging phong shading.
.IP "\fB-bouncedebug\fP"
Write bounced lighting only to the lightmap for debugging / previewing -bounce.
.IP "\fB-surflight_dump\fP"
Saves the lights generated by surfacelights to a "mapname-surflights.map" file.
.IP "\fB-novisapprox\fP"
Disable approximate visibility culling of lights, which has a small chance of introducing artifacts where lights cut off too soon.
.br
.SS "Experimental options:"
.IP "\fB-addmin\fP"
Changes the behaviour of \fIminlight\fP.  Instead of increasing low
light levels to the global minimum, add the global minimum light level
to all style 0 lightmaps.  This may help reducing the sometimes
uniform minlight effect.
.IP "\fB-lit2\fP"
Force generation of a .lit2 file, even if your map does not have any coloured
lights.
.IP "\fB-lux\fP"
Generate a .lux file storing average incoming light directions for surfaces. Usable by FTEQW with "r_deluxemapping 1"
.IP "\fB-lmscale n\fP"
Equivalent to "_lightmap_scale" worldspawn key.
.IP "\fB-bspxlit\fP"
Writes rgb data into the bsp itself.
.IP "\fB-bspx\fP"
Writes both rgb and directions data into the bsp itself.
.IP "\fB-novanilla\fP
Fallback scaled lighting will be omitted. Standard grey lighting will be omitted if there are coloured lights. Implies "-bspxlit". "-lit" will no longer be implied by the presence of coloured lights.

.SH "MODEL ENTITY KEYS"

.SS "Worldspawn Keys"

.PP
The following keys can be added to the \fIworldspawn\fP entity:

.IP "\fB""light"" ""n""\fP | \fB""_minlight"" ""n""\fP"
Set a global minimum light level of "n" across the whole map.  This is an easy
way to eliminate completely dark areas of the level, however you may lose some
contrast as a result, so use with care. Default 0.

.IP "\fB""_minlight_color"" ""r g b""\fP | \fB""_mincolor"" ""r g b""\fP"
Specify red(r), green(g) and blue(b) components for the colour of the
minlight. RGB component values are between 0 and 255 (between 0 and 1 is also
accepted). Default is white light ("255 255 255").

.IP "\fB""_dist"" ""n""\fP"
Scales the fade distance of all lights by a factor of n.  If n > 1 lights fade
more quickly with distance and if n < 1, lights fade more slowly with distance
and light reaches further.

.IP "\fB""_range"" ""n""\fP"
Scales the brightness range of all lights without affecting their fade
discance.  Values of n > 0.5 makes lights brighter and n < 0.5 makes lights
less bright.  The same effect can be achieved on individual lights by
adjusting both the "light" and "wait" attributes.

.IP "\fB""_sunlight"" ""n""\fP"
Set the brightness of the sunlight coming from an unseen sun in the sky.  Sky
brushes (or more accurately bsp leafs with sky contents) will emit sunlight at
an angle specified by the "_sun_mangle" key.  Default 0.

.IP "\fB""_anglescale"" ""n""\fP | \fB""_anglesense"" ""n""\fP"
Set the scaling of sunlight brightness due to the angle of incidence with a
surface (more detailed explanation in the "_anglescale" light entity key
below).

.IP "\fB""_sunlight_mangle"" ""yaw pitch roll""\fP | \fB""_sun_mangle"" ""yaw pitch roll""\fP"
Specifies the direction of sunlight using yaw, pitch and roll in
degrees. Yaw specifies the angle around the Z-axis from 0 to 359 degrees and
pitch specifies the angle from 90 (shining straight up) to -90 (shining straight down from above). Roll
has no effect, so use any value (e.g. 0).  Default is straight down ("0 -90
0").

.IP "\fB""_sunlight_penumbra"" ""n""\fP"
Specifies the penumbra width, in degrees, of sunlight.
Useful values are 3-4 for a gentle soft edge, or 10-20+ for more diffuse
sunlight. Default is 0.

.IP "\fB""_sunlight_color"" ""r g b""\fP"
Specify red(r), green(g) and blue(b) components for the colour of the
sunlight. RGB component values are between 0 and 255 (between 0 and 1 is also
accepted). Default is white light
("255 255 255").

.IP "\fB""_sunlight2"" ""n""\fP"
Set the brightness of a dome of lights arranged around the upper hemisphere.
(i.e. ambient light, coming from above the horizon). Default 0.

.IP "\fB""_sunlight_color2"" ""r g b""\fP | \fB""_sunlight2_color"" ""r g b""\fP"
Specifies the colour of _sunlight2, same format as "_sunlight_color". Default is
white light ("255 255 255").

.IP "\fB""_sunlight3"" ""n""\fP"
Same as "_sunlight2", but for the bottom hemisphere (i.e. ambient light, coming
from below the horizon). Combine "_sunlight2" and "_sunlight3" to have light coming equally
from all directions, e.g. for levels floating in the clouds. Default 0.

.IP "\fB""_sunlight_color3"" ""r g b""\fP | \fB""_sunlight3_color"" ""r g b""\fP"
Specifies the colour of "_sunlight3". Default is white light ("255 255 255").

.IP "\fB""_dirt"" ""n""\fP"
1 enables dirtmapping (ambient occlusion) on all lights, borrowed from q3map2. This adds shadows
to corners and crevices. You can override the global setting for specific lights with the
"_dirt" light entity key or "_sunlight_dirt", "_sunlight2_dirt", and "_minlight_dirt" worldspawn keys.
Default is no dirtmapping (-1).

.IP "\fB""_sunlight_dirt"" ""n""\fP"
1 enables dirtmapping (ambient occlusion) on sunlight, -1 to disable (making it illuminate the dirtmapping shadows). Default is to use the value of "_dirt".

.IP "\fB""_sunlight2_dirt"" ""n""\fP"
1 enables dirtmapping (ambient occlusion) on sunlight2/3, -1 to disable. Default is to use the value of "_dirt".

.IP "\fB""_minlight_dirt"" ""n""\fP"
1 enables dirtmapping (ambient occlusion) on minlight, -1 to disable. Default is to use the value of "_dirt".

.IP "\fB""_dirtmode"" ""n""\fP"
Choose between ordered (0, default) and randomized (1) dirtmapping.

.IP "\fB""_dirtdepth"" ""n""\fP"
Maximum depth of occlusion checking for dirtmapping, default 128.

.IP "\fB""_dirtscale"" ""n""\fP"
Scale factor used in dirt calculations, default 1. Lower values (e.g. 0.5) make
the dirt fainter, 2.0 would create much darker shadows.

.IP "\fB""_dirtgain"" ""n""\fP"
Exponent used in dirt calculation, default 1. Lower values (e.g. 0.5) make the
shadows darker and stretch further away from corners.

.IP "\fB""_dirtangle"" ""n""\fP"
Cone angle in degrees for occlusion testing, default 88. Allowed range 1-90.
Lower values can avoid unwanted dirt on arches, pipe interiors, etc.

.IP "\fB""_gamma"" ""n""\fP"
Adjust brightness of final lightmap. Default 1, >1 is brighter, <1 is darker.

.IP "\fB""_lightmap_scale"" ""n""\fP"
Forces all surfaces+submodels to use this specific lightmap scale. Removes "LMSHIFT" field.

.IP "\fB""_bounce"" ""n""\fP"
1 enables bounce lighting, disabled by default.

.IP "\fB""_bouncescale"" ""n""\fP"
Scales brightness of bounce lighting, default 1.

.IP "\fB""_bouncecolorscale"" ""n""\fP"
Weight for bounce lighting to use texture colors from the map: 0=ignore map textures (default), 1=multiply bounce light color by texture color.

.IP "\fB""_bouncestyled"" ""n""\fP"
1 makes styled lights bounce (e.g. flickering or switchable lights), default is 0, they do not bounce.

.IP "\fB""_spotlightautofalloff"" ""n""\fP"
When set to 1, spotlight falloff is calculated from the distance to the targeted info_null. Ignored when "_falloff" is not 0. Default 0.


.SS "Model Entity Keys"

.PP
The following keys can be used on any entity with a brush model.
"_minlight", "_mincolor", "_dirt", "_phong", "_phong_angle", "_phong_angle_concave", "_shadow", "_bounce" are supported on func_detail/func_group as well, if
qbsp from these tools is used.

.IP "\fB""_minlight"" ""n""\fP"
Set the minimum light level for any surface of the brush model.  Default 0.

.IP "\fB""_minlight_exclude"" ""texname""\fP"
Faces with the given texture are excluded from receiving minlight on this brush model.

.IP "\fB""_minlight_color"" ""r g b""\fP | \fB""_mincolor"" ""r g b""\fP"
Specify red(r), green(g) and blue(b) components for the colour of the
minlight. RGB component values are between 0 and 255 (between 0 and 1 is also
accepted). Default is white light
("255 255 255").

.IP "\fB""_shadow"" ""n""\fP"
If n is 1, this model will cast shadows on other models and itself
(i.e. "_shadow" implies "_shadowself").  Note that this doesn't magically give
Quake dynamic lighting powers, so the shadows will not move if the model
moves. Set to -1 on func_detail/func_group to prevent them from casting shadows. Default 0.

.IP "\fB""_shadowself"" ""n""\fP | \fB""_selfshadow"" ""n""\fP"
If n is 1, this model will cast shadows on itself if one part of the model
blocks the light from another model surface. This can be a better compromise
for moving models than full shadowing.  Default 0.

.IP "\fB""_shadowworldonly"" ""n""\fP"
If n is 1, this model will cast shadows on the world only (not other bmodels).

.IP "\fB""_switchableshadow"" ""n""\fP"
If n is 1, this model casts a shadow that can be switched on/off using QuakeC.
To make this work, a lightstyle is automatically assigned and stored in a key called "switchshadstyle",
which the QuakeC will need to read and call the "lightstyle()" builtin with "a" or "m" to switch the shadow on or off.
Entities sharing the same targetname, and with "_switchableshadow" set to 1, will share the same lightstyle.

These models are only able to block style 0 light (i.e., non-flickering or switchable lights).
Flickering or switchable lights will shine through the switchable shadow casters, regardless of whether the
shadow is off or on.

.IP "\fB""_dirt"" ""n""\fP"
For brush models, -1 prevents dirtmapping on the brush model. Useful if the
bmodel touches or sticks into the world, and you want to prevent those areas from
turning black. Default 0.

.IP "\fB""_phong"" ""n""\fP"
1 enables phong shading on this model with a default _phong_angle of 89 (softens columns etc).

.IP "\fB""_phong_angle"" ""n""\fP"
Enables phong shading on faces of this model with a custom angle. Adjacent faces with normals this many degrees apart (or less) will be smoothed.
Consider setting "_anglescale" to "1" on lights or worldspawn to make the effect of phong shading more visible.
Use the "-phongdebug" command-line flag to save the interpolated normals to the lightmap for previewing (use "r_lightmap 1" or "gl_lightmaps 1" in your engine to preview.)

.IP "\fB""_phong_angle_concave"" ""n""\fP"
Optional key for setting a different angle threshold for concave joints.
A pair of faces will either use "_phong_angle" or "_phong_angle_concave" as the smoothing threshold, depending on whether the joint between the faces is concave or not.
"_phong_angle(_concave)" is the maximum angle (in degrees) between the face normals that will still cause the pair of faces to be smoothed.
The minimum setting for "_phong_angle_concave" is 1, this should make all concave joints non-smoothed (unless they're less than 1 degree apart, almost a flat plane.)
If it's 0 or unset, the same value as "_phong_angle" is used.

.IP "\fB""_lightignore"" ""n""\fP"
1 makes a model receive minlight only, ignoring all lights / sunlight. Could be useful on rotators / trains.

.IP "\fB""_bounce"" ""n""\fP"
Set to -1 to prevent this model from bouncing light (i.e. prevents its brushes from emitting bounced light they receive from elsewhere.) Only has an effect if "_bounce" is enabled in worldspawn.



.SH "LIGHT ENTITY KEYS"

.PP
Light entity keys can be used in any entity with a classname starting
with the first five letters "light". E.g. "light", "light_globe",
"light_flame_small_yellow", etc.

.IP "\fB""light"" ""n""\fP"
Set the light intensity. Negative values are also allowed and will cause the
entity to subtract light cast by other entities. Default 300.

.IP "\fB""wait"" ""n""\fP"
Scale the fade distance of the light by "n". Values of n > 1 make the light
fade more quickly with distance, and values < 1 make the light fade more
slowly (and thus reach further). Default 1.

.IP "\fB""delay"" ""n""\fP"
Select an attenuation formaula for the light:
.nf
  0 => Linear attenuation (default)
  1 => 1/x attenuation
  2 => 1/(x^2) attenuation
  3 => No attenuation (same brightness at any distance)
  4 => "local minlight" - No attenuation and like minlight,
       it won't raise the lighting above it's light value.
       Unlike minlight, it will only affect surfaces within
       line of sight of the entity.
  5 => 1/(x^2) attenuation, but slightly more attenuated and
       without the extra bright effect that "delay 2" has
       near the source.
.fi

.IP "\fB""_falloff"" ""n""\fP"
Sets the distance at which the light drops to 0, in map units.

In this mode, "wait" is ignored and "light" only controls the brightness at the center
of the light, and no longer affects the falloff distance.

Only supported on linear attenuation (delay 0) lights currently.

.IP "\fB""_color"" ""r g b""\fP"
Specify red(r), green(g) and blue(b) components for the colour of the
light. RGB component values are between 0 and 255 (between 0 and 1 is also
accepted). Default is white light
("255 255 255").

.IP "\fB""target"" ""name""\fP"
Turns the light into a spotlight, with the direction of light being towards
another entity with it's "targetname" key set to "name".

.IP "\fB""mangle"" ""yaw pitch roll""\fP"
Turns the light into a spotlight and specifies the direction of light using
yaw, pitch and roll in degrees. Yaw specifies the angle around the
Z-axis from 0 to 359 degrees and pitch specifies the angle from 90 (straight
up) to -90 (straight down). Roll has no effect, so use any value (e.g. 0).
Often easier than the "target" method.

.IP "\fB""angle"" ""n""\fP"
Specifies the angle in degrees for a spotlight cone. Default 40.

.IP "\fB""_softangle"" ""n""\fP"
Specifies the angle in degrees for an inner spotlight cone (must be less than
the "angle" cone. Creates a softer transition between the full brightness of
the inner cone to the edge of the outer cone.  Default 0 (disabled).

.IP "\fB""targetname"" ""name""\fP"
Turns the light into a switchable light, toggled by another entity targeting
it's name.

.IP "\fB""style"" ""n""\fP"
Set the animated light style. Default 0.

.IP "\fB""_anglescale"" ""n""\fP | \fB""_anglesense"" ""n""\fP"
Sets a scaling factor for how much influence the angle of incidence of light
on a surface has on the brightness of the surface. \fIn\fP must be between 0.0
and 1.0. Smaller values mean less attenuation, with zero meaning that angle of
incidence has no effect at all on the brightness. Default 0.5.

.IP "\fB""_dirtscale"" ""n""\fP | \fB""_dirtgain"" ""n""\fP"
Override the global "_dirtscale" or "_dirtgain" settings to change how this
light is affected by dirtmapping (ambient occlusion). See descriptions of these
keys in the worldspawn section.

.IP "\fB""_dirt"" ""n""\fP"

tyrquake-0.70/changelog.txt
/-----------\
  CHANGELOG
\-----------/

v0.70
=====
- Fix 'Cache_Dealloc:: not allocated' error on invalid QW skins
- Fix some translucent alias model rendering bugs in the x86 assembly
- Fix some invalid GL api usage (thanks Jelle Raaijmakers)

v0.69
=====
- Add support for scaling the software renderer output by a factor of 2 or 4
- Add translucent HUD support ("scr_sbaralpha")
- Add support for OpenGL vertex program animation interpolation
- Add support for OpenGL fragment program water surface warping
- Add support for OpenGL buffer objects to improve rendering performance
- Add support for DXT texture compression (greatly helps low-memory systems)
- Add vsync support for most platforms ("vid_vsync")
- Add anisotropic filtering for OpenGL ("gl_texture_anisotropy")
- Add back the vid mode test functionality
- Add cl_maxpitch and cl_minpitch console variables (default 90/-90)
- Improve FOV calculations to work better with widescreen aspect ratios
- Improved dynamic lighting performance on OpenGL (removed redundant updates)
- Fix a few software renderer entity alpha bugs
- Fix a few more texture reloading bugs in OpenGL
- Fix framerate on windows when compiled with SDL video/input
- Fix a few model/sprite clipping bugs the could crash the software renderer

v0.68
=====
- Fix compilation of x86 on unix variant when using ASM
- Fix the assembly patch process for translucent alias models
- Fix crash caused by missing flush of overflow sfx on level change
- Fix crash caused by missing flush of overflow models on level change
- Fix gl texture leak when alias models are pushed out of the cache due
  to memory pressure.
- Make the default zone size relative to the overall memory size

v.067
=====
- Added (finally) loading new mods at runtime with the 'game' command.
- Added ability to play music tracks from sound files.  Currently ogg
  (vorbis), mp3, flac and wav are supported.
- Add some auto-detection of quake installation directory (windows).
  If the game is started without specifying the base directory and the
  working directory is not a valid basedir, the game will try to check
  if you have a copy of Quake in Steam and use that.
- Improved console tab completion for commands with multiple arguments
- Fixed overflowing the sound effects cache
- Improved detection of local network address
- Improved reliability of game discovery via broadcast on LANs
- Fixed fullbright textures on player skins
- Added a pulseaudio sound driver, now the default on Linux.
- Fixes for minor audio artifacts in looping sounds
- Fix crash in NQ dedicated server
- Higher quality (resolution) underwater warping in software renderer
- Fix particle sizes in software and GL
- Make a fixed size LRU cache of textures in GLQuake and free unused textures

v0.66
=====
- Reworked video mode setting, now uses vid_width, vid_height, etc. cvars
- Added menu for manually specifying windowed mode resolution
- Improved window focus handling with respect to mouse/keyboard grabs
- Improved window icon handling in SDL/X11 environments
- Implement improved quality water warp effect from FQ and others.
- Fixed silly bug which was making OpenGL mode changes stupidly slow
- Fixed crash that could happen when a BSP has missing textures

v0.65
=====
- Fix crash in the software renderer
- Fix transparent menupics on the video options menu

v0.64
=====
- Entity alpha support via FitzQuake protocol
- FitzQuake protocol (666) is now the default for NQ
- Coloured lighting, fog and skybox support in the OpenGL renderer
- Alpha blending (water/slime/lava/tele + entities) (GL + SW)
- Fence texture support (GL + SW)
- Fixed non-64x64 turbulent surfaces in the software renderer
- Improved sky in GL (from FitzQuake/Quakespasm)
- Alias model lighting interpolation (from Darkplaces)
- Automatic adjustment of r_maxsurfs / r_maxedges in the software renderer
- Model interpolation now works in QW as well

v0.63
=====
- OpenGL renderer improvements:
  - Much improved speed by batching drawing operations
  - Fullbright pixels now properly rendered
  - If 3+ texture units are available, all materials can be drawn in a single pass
  - Improved fidelity of model skins when NPoT textures unavailable
- Fix some of the rough edges on mouse grab handling
- Fixed bug where red/alt console text was not being rendered
- New 'hudscale' command to scale the size of HUD elements (status bar, etc.)
- Fixed crosshair rendering for 'crosshair 2'
- Fixed custom QW skin loading on PPC
- Updated build system to enable a 'legacy' Universal build on OSX 10.5 and
    an x64-only build on modern OSX systems.
- Added a custom Icon for Windows/OSX builds

v0.62
=====
- Model animation/movement interpolation (r_lerpmodels, r_lerpmove cvars)
- SDL video/input/sound target, now default for MacOS instead of X11.
- sndio sound target for OpenBSD (thanks Tony Spets/Jacob Meuser)
- Support for BJP1/2/3 and FITZ(666) network/demo protocols
- Support for Darkplaces and RMQ style BSP2/2PSB file formats
- Work around buggy timers on some windows systems/drivers (thanks MH!)
- Removed vid_win dependency on SciTech MGL (thanks MH!)
- Netmasks can be used with qwsv's addip/removeip; e.g. "addip 192.168.0.0/24"
- Fixed remote screenshots from QWSV
- Improved build system including new OSX, Win64, SDL targets
- Lots more code sharing between NQ/QW
- Use non-power-of-two OpenGL textures when extension is available
- New cvar "m_freelook" to replace need for manual editing "+mlook" in config
- Improvements to handling of GL textures
- Support changing video mode in GLQuake/GLQWCL (finally!)
- Ten years since v0.61!

v0.61
=====
- Fix QWSV command line parsing
- Attempt to detect when X86 assembler files should be used

v0.60
=====
- Fix video buffer overrun when rendering endtitle in low-res vid mode
- Reduce the load timeout when a changelevel fails

v0.59
=====
- Various improvements to the windows video code
  - Start in windowed mode to avoid extra mode changes on the monitor

tyrutils-0.17/light/light.c
/*  Copyright (C) 1996-1997  Id Software, Inc.

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

    See file, 'COPYING', for details.
*/

#include <stdint.h>

#include <light/light.h>
#include <light/entities.h>

float scaledist = 1.0;
float rangescale = 0.5;
float anglescale = 0.5;
float sun_anglescale = 0.5;
float fadegate = EQUAL_EPSILON;
int softsamples = 0;
const vec3_t vec3_white = { 255, 255, 255 };

qboolean addminlight = false;
lightsample_t minlight = { 0, { 255, 255, 255 } };
lightsample_t sunlight = { 0, { 255, 255, 255 } };
vec3_t sunvec = { 0, 0, 16384 };		/* defaults to straight down */

byte *filebase;			// start of lightmap data
static byte *file_p;		// start of free space after data
static byte *file_end;		// end of free space for lightmap data

byte *lit_filebase;		// start of litfile data
static byte *lit_file_p;	// start of free space after litfile data
static byte *lit_file_end;	// end of space for litfile data

static modelinfo_t *modelinfo;
const dmodel_t *const *tracelist;

int oversample = 1;
qboolean write_litfile = false;
qboolean litfile_only = false;

int32_t
AllocateLightDataOffset(int numsamples)
{
    int32_t offset;

    ThreadLock();

    /* align to 4 byte boudaries */
    file_p = (byte *)(((uintptr_t)file_p + 3) & ~3);
    offset = file_p - filebase;

    file_p += numsamples;

    /* align to 12 byte boundaries to match offets with 3 * lightdata */
    if ((uintptr_t)lit_file_p % 12)
        lit_file_p += 12 - ((uintptr_t)lit_file_p % 12);
    lit_file_p += numsamples * 3;

    ThreadUnlock();

    if (file_p > file_end)
	Error("%s: overrun", __func__);

    if (lit_file_p > lit_file_end)
	Error("%s: overrun", __func__);

    return offset;
}

byte *
LightDataDest(const bsp2_dface_t *face)
{
    return filebase + face->lightofs;
}

byte *
ColorDataDest(const bsp2_dface_t *face)
{
    return lit_filebase + 3 * face->lightofs;
}

static void *
LightThread(void *arg)
{
    int facenum, i;
    dmodel_t *model;
    const bsp2_t *bsp = arg;

    while (1) {
	facenum = GetThreadWork();
	if (facenum == -1)
	    break;

	/* Find the correct model offset */
	for (i = 0, model = bsp->dmodels; i < bsp->nummodels; i++, model++) {
	    if (facenum < model->firstface)
		continue;
	    if (facenum < model->firstface + model->numfaces)
		break;
	}
	if (i == bsp->nummodels) {
	    logprint("warning: no model has face %d\n", facenum);
	    continue;
	}

	LightFace(bsp->dfaces + facenum, &modelinfo[i], bsp);
    }

    return NULL;
}

static void
FindModelInfo(const bsp2_t *bsp)
{
    int i, shadow, numshadowmodels;
    entity_t *entity;
    char modelname[20];
    const char *attribute;
    const dmodel_t **shadowmodels;
    modelinfo_t *info;

    shadowmodels = malloc(sizeof(dmodel_t *) * (bsp->nummodels + 1));
    memset(shadowmodels, 0, sizeof(dmodel_t *) * (bsp->nummodels + 1));

    /* The world always casts shadows */
    shadowmodels[0] = &bsp->dmodels[0];
    numshadowmodels = 1;

    memset(modelinfo, 0, sizeof(*modelinfo) * bsp->nummodels);
    modelinfo[0].model = &bsp->dmodels[0];

    for (i = 1, info = modelinfo + 1; i < bsp->nummodels; i++, info++) {
	info->model = &bsp->dmodels[i];

	/* Find the entity for the model */
	snprintf(modelname, sizeof(modelname), "*%d", i);
	entity = FindEntityWithKeyPair("model", modelname);
	if (!entity)
	    Error("%s: Couldn't find entity for model %s.\n", __func__,
		  modelname);

	/* Check if this model will cast shadows (shadow => shadowself) */
	shadow = atoi(ValueForKey(entity, "_shadow"));
	if (shadow) {
	    shadowmodels[numshadowmodels++] = &bsp->dmodels[i];
	} else {
	    shadow = atoi(ValueForKey(entity, "_shadowself"));
	    if (shadow)
		info->shadowself = true;
	}

	/* Set up the offset for rotate_* entities */
	attribute = ValueForKey(entity, "classname");
	if (!strncmp(attribute, "rotate_", 7))
	    GetVectorForKey(entity, "origin", info->offset);

	/* Grab the bmodel minlight values, if any */
	attribute = ValueForKey(entity, "_minlight");
	if (attribute[0])
	    info->minlight.light = atoi(attribute);
	GetVectorForKey(entity, "_mincolor", info->minlight.color);
	if (!VectorCompare(info->minlight.color, vec3_origin)) {
	    if (!write_litfile)
		write_litfile = true;
	} else {
	    VectorCopy(vec3_white, info->minlight.color);
	}
    }

    tracelist = shadowmodels;
}

/*
 * =============
 *  LightWorld
 * =============
 */
static void
LightWorld(bsp2_t *bsp)
{
    if (bsp->dlightdata)
	free(bsp->dlightdata);

    if (!litfile_only)
        bsp->lightdatasize = MAX_MAP_LIGHTING; /* FIXME - remove this limit */

    bsp->dlightdata = malloc(bsp->lightdatasize * 4 + 16); /* for alignment */
    if (!bsp->dlightdata)
	Error("%s: allocation of %i bytes failed.", __func__, bsp->lightdatasize * 4);
    memset(bsp->dlightdata, 0, bsp->lightdatasize * 4 + 16);

    /* align filebase to a 4 byte boundary */
    filebase = file_p = (byte *)(((uintptr_t)bsp->dlightdata + 3) & ~3);
    file_end = filebase + bsp->lightdatasize;

    /* litfile data stored in dlightdata, after the white light */
    lit_filebase = file_end + 12 - ((uintptr_t)file_end % 12);
    lit_file_p = lit_filebase;
    lit_file_end = lit_filebase + 3 * bsp->lightdatasize;

    RunThreadsOn(0, bsp->numfaces, LightThread, bsp);
    logprint("Lighting Completed.\n\n");

    if (!litfile_only)
        bsp->lightdatasize = file_p - filebase;
    logprint("lightdatasize: %i\n", bsp->lightdatasize);
}


/*
 * ==================
 * main
 * light modelfile
 * ==================
 */
int
main(int argc, const char **argv)
{
    bspdata_t bspdata;
    bsp2_t *const bsp = &bspdata.data.bsp2;
    int32_t loadversion;
    int i;
    double start;
    double end;
    char source[1024];

    init_log("light.log");
    logprint("---- light / TyrUtils " stringify(TYRUTILS_VERSION) " ----\n");

    numthreads = GetDefaultThreads();

    for (i = 1; i < argc; i++) {
	if (!strcmp(argv[i], "-threads")) {
	    numthreads = atoi(argv[++i]);
	} else if (!strcmp(argv[i], "-extra")) {
	    oversample = 2;
	    logprint("extra 2x2 sampling enabled\n");
	} else if (!strcmp(argv[i], "-extra4")) {
	    oversample = 4;
	    logprint("extra 4x4 sampling enabled\n");
	} else if (!strcmp(argv[i], "-dist")) {
	    scaledist = atof(argv[++i]);
	} else if (!strcmp(argv[i], "-range")) {
	    rangescale = atof(argv[++i]);
	} else if (!strcmp(argv[i], "-gate")) {
	    fadegate = atof(argv[++i]);
	} else if (!strcmp(argv[i], "-light")) {
	    minlight.light = atof(argv[++i]);
	} else if (!strcmp(argv[i], "-addmin")) {
	    addminlight = true;
	} else if (!strcmp(argv[i], "-lit")) {
	    write_litfile = true;
	} else if (!strcmp(argv[i], "-litonly")) {
            write_litfile = true;
	    litfile_only = true;
	} else if (!strcmp(argv[i], "-soft")) {
	    if (i < argc - 2 && isdigit(argv[i + 1][0]))
		softsamples = atoi(argv[++i]);
	    else
		softsamples = -1; /* auto, based on oversampling */
	} else if (!strcmp(argv[i], "-anglescale") || !strcmp(argv[i], "-anglesense")) {
	    if (i < argc - 2 && isdigit(argv[i + 1][0]))
		anglescale = sun_anglescale = atoi(argv[++i]);
	    else
		Error("-anglesense requires a numeric argument (0.0 - 1.0)");
	} else if (argv[i][0] == '-')
	    Error("Unknown option \"%s\"", argv[i]);
	else
	    break;
    }

    if (i != argc - 1) {
	printf("usage: light [-threads num] [-extra|-extra4]\n"
	       "             [-light num] [-addmin] [-anglescale|-anglesense]\n"
	       "             [-dist n] [-range n] [-gate n] [-lit] [-litonly]\n"
	       "             [-soft [n]] bspfile\n");
	exit(1);
    }

    if (numthreads > 1)
	logprint("running with %d threads\n", numthreads);
    if (write_litfile)
	logprint(".lit colored light output requested on command line.\n");
    if (softsamples == -1) {
	switch (oversample) {
	case 2:
	    softsamples = 1;
	    break;
	case 4:
	    softsamples = 2;
	    break;
	default:
	    softsamples = 0;
	    break;
	}
    }

    start = I_FloatTime();

    strcpy(source, argv[i]);
    StripExtension(source);
    DefaultExtension(source, ".bsp");
    LoadBSPFile(source, &bspdata);

    loadversion = bspdata.version;
    if (bspdata.version != BSP2VERSION)
	ConvertBSPFormat(BSP2VERSION, &bspdata);

    LoadEntities(bsp);
    MakeTnodes(bsp);
    modelinfo = malloc(bsp->nummodels * sizeof(*modelinfo));
    FindModelInfo(bsp);
    LightWorld(bsp);
    free(modelinfo);

    WriteEntitiesToString(bsp);

    if (write_litfile)
	WriteLitFile(bsp, source, LIT_VERSION);

    /* Convert data format back if and write out if necessary */
    if (!litfile_only) {
        if (loadversion != BSP2VERSION)
            ConvertBSPFormat(loadversion, &bspdata);

        WriteBSPFile(source, &bspdata);
    }

    end = I_FloatTime();
    logprint("%5.1f seconds elapsed\n", end - start);

    close_log();

    return 0;
}
